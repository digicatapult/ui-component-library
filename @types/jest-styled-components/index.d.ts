declare module 'jest-styled-components/serializer'
