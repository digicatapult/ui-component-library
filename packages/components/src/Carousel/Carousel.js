import React, { useState } from 'react'
import { PropTypes } from 'prop-types'
import styled from 'styled-components'
import Button from '../Button'

const Wrapper = styled.div`
  position: relative;
  display: flex;
  justify-content: space-between;
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  overflow: hidden;
`

const ImagesWrapper = styled.div`
  position: absolute;
  display: flex;
  flex-direction: row;
  flex-wrap: no-wrap;
  transform: translate(
    -${({ activeIndex, numberOfImages }) => (activeIndex * 100) / numberOfImages}%
  );
  transition: transform 0.5s;
`

const Image = styled.img`
  object-fit: cover;
  width: ${({ width }) => width};
  height: ${({ height }) => height};
`

const StyledButton = styled(Button)`
  z-index: 1;
  :disabled {
    background-color: inherit;
    color: #808080;
  }
`

const Images = ({ height, width, activeIndex, images }) => (
  <ImagesWrapper
    numberOfImages={images.length}
    width={width}
    activeIndex={activeIndex}
  >
    {images.map((image, idx) => (
      <div key={idx}>
        <Image height={height} width={width} alt={image.alt} src={image.src} />
      </div>
    ))}
  </ImagesWrapper>
)

const Carousel = ({ images, height, width, ...rest }) => {
  const [activeIndex, setActiveIndex] = useState(0)

  const handleNavigation = e => {
    const leftKeyCode = 37
    const rightKeyCode = 39

    if (e.target.name === 'previous' || e.keyCode === leftKeyCode) {
      setActiveIndex(Math.max(activeIndex - 1, 0))
    } else if (e.target.name === 'next' || e.keyCode === rightKeyCode) {
      setActiveIndex(Math.min(activeIndex + 1, images.length - 1))
    }
  }

  return (
    <Wrapper
      height={height}
      width={width}
      onKeyDown={handleNavigation}
      tabIndex="0"
      {...rest}
    >
      <StyledButton
        left
        name="previous"
        disabled={activeIndex === 0}
        onClick={handleNavigation}
        type="button"
        background="linear-gradient(to left, #00000000, #00000080)"
      >
        prev
      </StyledButton>
      <Images
        height={height}
        width={width}
        images={images}
        activeIndex={activeIndex}
      />
      <StyledButton
        right
        name="next"
        disabled={activeIndex === images.length - 1}
        onClick={handleNavigation}
        type="button"
        background="linear-gradient(to right, #00000000, #00000080)"
      >
        next
      </StyledButton>
    </Wrapper>
  )
}

Images.defaultProps = {
  activeIndex: 0,
  ...Carousel.defaultProps
}

Images.propTypes = {
  activeIndex: PropTypes.number,
  ...Carousel.propTypes
}

Carousel.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string,
      alt: PropTypes.string
    })
  ).isRequired,
  height: PropTypes.string,
  width: PropTypes.string
}

Carousel.defaultProps = {
  height: '90vh',
  width: '90vw'
}

export default Carousel
