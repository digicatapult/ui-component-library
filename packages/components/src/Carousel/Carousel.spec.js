import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Carousel from '.'

const images = [
  {
    src: '/mock/path/1',
    alt: 'first-mock'
  },
  {
    src: 'mock/path/2',
    alt: 'second-mock'
  }
]

describe('Carousel', () => {
  test('Image Carousel', () => {
    const tree = renderer
      .create(<Carousel width="90vw" height="90vh" images={images} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
