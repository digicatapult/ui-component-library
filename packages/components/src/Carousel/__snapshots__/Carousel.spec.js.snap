// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Image Carousel 1`] = `
.c1 {
  padding: 16px;
  color: #fff;
  border-radius: 4px;
  background: linear-gradient(to left,#00000000,#00000080);
  border: none;
  box-sizing: border-box;
  text-transform: uppercase;
  cursor: pointer;
}

.c1:hover {
  outline: none;
  -webkit-filter: brightness(120%);
  filter: brightness(120%);
}

.c1:focus {
  outline: none;
}

.c1:disabled {
  cursor: default;
  background-color: #dfdfdf;
  -webkit-filter: none;
  filter: none;
  color: #fff;
  border: none;
}

.c5 {
  padding: 16px;
  color: #fff;
  border-radius: 4px;
  background: linear-gradient(to right,#00000000,#00000080);
  border: none;
  box-sizing: border-box;
  text-transform: uppercase;
  cursor: pointer;
}

.c5:hover {
  outline: none;
  -webkit-filter: brightness(120%);
  filter: brightness(120%);
}

.c5:focus {
  outline: none;
}

.c5:disabled {
  cursor: default;
  background-color: #dfdfdf;
  -webkit-filter: none;
  filter: none;
  color: #fff;
  border: none;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 90vw;
  height: 90vh;
  overflow: hidden;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: no-wrap;
  -ms-flex-wrap: no-wrap;
  flex-wrap: no-wrap;
  -webkit-transform: translate( -0% );
  -ms-transform: translate( -0% );
  transform: translate( -0% );
  -webkit-transition: -webkit-transform 0.5s;
  -webkit-transition: transform 0.5s;
  transition: transform 0.5s;
}

.c4 {
  object-fit: cover;
  width: 90vw;
  height: 90vh;
}

.c2 {
  z-index: 1;
}

.c2:disabled {
  background-color: inherit;
  color: #808080;
}

<div
  className="c0"
  height="90vh"
  onKeyDown={[Function]}
  tabIndex="0"
  width="90vw"
>
  <button
    className="c1 c2"
    disabled={true}
    name="previous"
    onClick={[Function]}
    size="medium"
    type="button"
  >
    prev
  </button>
  <div
    className="c3"
    width="90vw"
  >
    <div>
      <img
        alt="first-mock"
        className="c4"
        height="90vh"
        src="/mock/path/1"
        width="90vw"
      />
    </div>
    <div>
      <img
        alt="second-mock"
        className="c4"
        height="90vh"
        src="mock/path/2"
        width="90vw"
      />
    </div>
  </div>
  <button
    className="c5 c2"
    disabled={false}
    name="next"
    onClick={[Function]}
    size="medium"
    type="button"
  >
    next
  </button>
</div>
`;
